service/controller/v2/resource/instance/create.go

```diff
@@ -107,17 +135,82 @@ func (r *Resource) ensureDeploymentUpdate(ctx context.Context, customObject prov
 
 		g := key.ResourceGroupName(customObject)
 		s := deploymentNameFunc(customObject)
-		ids := compute.VirtualMachineScaleSetVMInstanceRequiredIDs{
+		ids := &compute.VirtualMachineScaleSetVMInstanceIDs{
 			InstanceIds: to.StringSlicePtr([]string{
-				instanceID,
+				*instanceToReimage.InstanceID,
 			}),
 		}
-		_, err = c.UpdateInstances(ctx, g, s, ids)
+		_, err = c.Reimage(ctx, g, s, ids)
```

Comment:
Another idea. Do by default "update" and do "reimage" only when we detect version bundle change?

Do not have separate loop for "reimage", but do it together (if necessary) with "update".
Smth like that?
```
	//     loop 1: worker 1 update (+ reimage if version change)
	//     loop 2: worker 2 update (+ reimage if version change)
	//     loop 3: worker 3 update (+ reimage if version change)
```

Reply:
I think we cannot do the calls after each other because the update itself only to apply the latest model takes quite long. I am not sure how ARM executes or guarantees that shit happens as we want it. We would need to test this but I would also not want to rely on too much magic. When we manage the process with different reconciliation loops and iterate gradually towards the desired state the system design is way more stable and reliable IMO. We can work on some different action scheduling in the future but as we discussed the suggested algorithm would be my first attempt to move forward.

Reply:
Are the newest tags applied to instances after after calling `update` or after vmss ARM template is deployed?

Reply:
> Right now the tags are not updated.

Please read what I wrote. I actually asked a question that I would like to get an answer to.

Reply:
I read it a few times already. I don't understand it fully. Does `Right now` mean before calling `Reimage` or before calling `Update` or something else?

Reply:
Well, what is the code doing? We call update, we call reimage. The tags are not updated. What is missing Pawel?

Reply:
Now I understand, tags aren't applied at at all. I understand they are applied to the vmss but not to the instances themselves. Maybe you can use the SDK to make the job done. I wouldn't be surprised if ARM removed the tags after next resync period though.

Reply:
Yes, this is the point. We have to manage the tags and I am not sure how. Need to get into this and see what we can do.

Reply:
Can you do reimage request immediately after update request?

Reply:
Uncertain. I would like to prevent this actually to let the system settle and have more control over it. I will test this if I have to though.